core/src/main/scala/org/bykn/bosatsu/SourceConverter.scala

```diff
@@ -423,24 +454,125 @@ final class SourceConverter(
+              def get(b: Bindable): Result[Pattern[(PackageName, Constructor), TypeRef]] =
+                mapping.get(b) match {
+                  case Some(pat) =>
+                    SourceConverter.success(pat)
+                  case None =>
+                    SourceConverter.partial(SourceConverter.MissingArg(nm, pat, present, b, region), Pattern.WildCard)
+                }
+              val mapped =
+                params
+                  .traverse { case (b, _) => get(b) }(SourceConverter.parallelIor)
+                  .map(Pattern.PositionalStruct(pc, _))
+
+              val paramNames = params.map(_._1).toSet
+              // here are all the fields we don't understand
+              val extra = fs.toList.iterator.map(_.field).filterNot(paramNames).toList
+              // Check that the mapping is exactly the right size
+              NonEmptyList.fromList(extra) match {
+                case None => mapped
+                case Some(extra) =>
+                  SourceConverter
+                    .addError(mapped,
+                      SourceConverter.UnexpectedField(nm, pat, extra, params.map(_._1), region))
+              }
+            case None =>
+              SourceConverter.failure(SourceConverter.UnknownConstructor(nm, pat, region))
+          }
+        }
+      case (Pattern.StructKind.NamedPartial(nm, Pattern.StructKind.Style.RecordLike(fs)), rargs) =>
+        rargs.flatMap { args =>
+          val pc@(p, c) = nameToCons(nm)
+          localTypeEnv.getConstructor(p, c) match {
+            case Some((params, _, _)) =>
+              val mapping = fs.toList.iterator.map(_.field).zip(args.iterator).toMap
+              def get(b: Bindable): Pattern[(PackageName, Constructor), TypeRef] =
+                mapping.get(b) match {
+                  case Some(pat) => pat
+                  case None => Pattern.WildCard
+                }
+              val derefArgs = params.map { case (b, _) => get(b) }
+              val res0 = SourceConverter.success(Pattern.PositionalStruct(pc, derefArgs))
+
+              val paramNames = params.map(_._1).toSet
+              // here are all the fields we don't understand
+              val extra = fs.toList.iterator.map(_.field).filterNot(paramNames).toList
+              // Check that the mapping is exactly the right size
+              NonEmptyList.fromList(extra) match {
+                case None => res0
+                case Some(extra) =>
+                  SourceConverter
+                    .addError(res0,
+                      SourceConverter.UnexpectedField(nm, pat, extra, params.map(_._1), region))
+              }
+            case None =>
+              SourceConverter.failure(SourceConverter.UnknownConstructor(nm, pat, region))
+          }
+        }
+    }(SourceConverter.parallelIor) // use the parallel, not the default Applicative which is Monadic
+    .map(_.mapType(toType))
```

Comment:
we should use a `Result` type in toType as well so we record any errors we get looking up the type.